webrtc/libjingle/xmpp/boshsocket.cc

```diff
@@ -0,0 +1,604 @@
+    output_queue_.pop();
+    auto temp_socket = primary_socket_;
+    if (temp_socket->socket_ != socket) {
+      temp_socket = secondary_socket_;
+    }
+    if (buffer_.Length() != 0) {
+      while (buffer_.Length() != 0) {
+        int written = socket->Send(buffer_.Data(), buffer_.Length());
+        if (written > 0) {
+          buffer_.Consume(written);
+          continue;
+        }
+        if (!socket->IsBlocking())
+          LOG(LS_ERROR) << "Send error: " << socket->GetError();
+        return;
+      }
+      temp_socket->pending_ = true;
+    }
+    last_send_ = time(NULL);
+  }
+
+  void BoshSocket::OnConnectEvent(rtc::AsyncSocket * socket) {
+    auto temp_socket = primary_socket_;
+    if (temp_socket->socket_ != socket) {
+      temp_socket = secondary_socket_;
+    }
+#if defined(FEATURE_ENABLE_SSL)
+    if (temp_socket->state_ == buzz::AsyncSocket::STATE_TLS_CONNECTING) {
+      temp_socket->state_ = buzz::AsyncSocket::STATE_TLS_OPEN;
+      temp_socket->connected_ = true;
+      temp_socket->pending_ = false;
+      SignalSSLConnected();
+      if (BothConnected()) {
+        OnWriteEvent(GetSocket()->socket_);
+      }
+      return;
+    }
+#endif  // !defined(FEATURE_ENABLE_SSL)
+    temp_socket->state_ = buzz::AsyncSocket::STATE_OPEN;
+    if (use_proxy_ && ssl_ && !temp_socket->connect_sent_) {
+      SendConnect(temp_socket);
+    }
+    else {
+      temp_socket->connected_ = true;
+      temp_socket->pending_ = false;
+      if (BothConnected()) {
+        SignalConnected();
+      }
+    }
+  }
+
+  void BoshSocket::OnCloseEvent(rtc::AsyncSocket * socket, int error) {
+    SignalCloseEvent(error);
+  }
+
+#else
+  void BoshSocket::OnEvent(rtc::StreamInterface* stream,
+                           int events, int err) {
+    auto temp_socket = primary_socket_;
+    if(temp_socket->stream_ != stream) {
+      temp_socket = secondary_socket_;
+    }
+
+    if ((events & rtc::SE_OPEN)) {
+#if defined(FEATURE_ENABLE_SSL)
+      if (temp_socket->state_ == buzz::AsyncSocket::STATE_TLS_CONNECTING) {
+        temp_socket->state_ = buzz::AsyncSocket::STATE_TLS_OPEN;
+        temp_socket->connected_ = true;
+        temp_socket->pending_ = false;
+        SignalSSLConnected();
+        if(BothConnected()) {
+          OnEvent(GetSocket()->stream_, rtc::SE_WRITE, 0);
+        }
+      }
+#endif
+      temp_socket->state_ = buzz::AsyncSocket::STATE_OPEN;
+      if(use_proxy_ && ssl_ && !temp_socket->connect_sent_) {
+        SendConnect(temp_socket);
+      }
+      else {
+        temp_socket->connected_ = true;
+        temp_socket->pending_ = false;
+        if (BothConnected()) {
+          SignalConnected();
+        }
+      }
+    }
+    if ((events & rtc::SE_READ)) {
+      last_receive_ = time(NULL);
+      buzz::InputMsg input_msg;
+      char bytes[4096];
+      size_t read;
+      rtc::StreamResult result = stream->Read(bytes, sizeof(bytes), &read, NULL);
```

Comment:
I've just update code to support building webrtc with the flag USE_SSLSTREAM enabled. But currently in Windows, we can establish HTTPS connection with ssl stream because webrtc currently don't support SChannel.